pkg/services/dashboards/service/zanzana.go

```diff
@@ -0,0 +1,325 @@
+	}
+
+	return first.result, first.err
+}
+
+func (dr *DashboardServiceImpl) findDashboardsZanzana(ctx context.Context, query *dashboards.FindPersistedDashboardsQuery) ([]dashboards.DashboardSearchProjection, error) {
+	findDashboards := dr.getFindDashboardsFn(query)
+	return findDashboards(ctx, query)
+}
+
+type findDashboardsFn func(ctx context.Context, query *dashboards.FindPersistedDashboardsQuery) ([]dashboards.DashboardSearchProjection, error)
+
+// getFindDashboardsFn makes a decision which search method should be used
+func (dr *DashboardServiceImpl) getFindDashboardsFn(query *dashboards.FindPersistedDashboardsQuery) findDashboardsFn {
+	if query.Limit > 0 && query.Limit < maxListQueryLimit && len(query.Title) > 0 {
+		return dr.findDashboardsZanzanaCheck
+	}
+	if len(query.DashboardUIDs) > 0 && len(query.DashboardUIDs) < maxListQueryLimit {
+		return dr.findDashboardsZanzanaCheck
+	}
+	if len(query.FolderUIDs) > 0 && len(query.FolderUIDs) < maxListQueryLimit {
+		return dr.findDashboardsZanzanaCheck
+	}
+	if len(query.Title) <= maxListQueryLength {
+		return dr.findDashboardsZanzanaList
+	}
+	return dr.findDashboardsZanzanaCheck
+}
+
+// findDashboardsZanzanaCheck implements "Search, then check" strategy. It first performs search query, then filters out results
+// by checking access to each item.
+func (dr *DashboardServiceImpl) findDashboardsZanzanaCheck(ctx context.Context, query *dashboards.FindPersistedDashboardsQuery) ([]dashboards.DashboardSearchProjection, error) {
+	ctx, span := tracer.Start(ctx, "dashboards.service.findDashboardsZanzanaCheck")
+	defer span.End()
+
+	result := make([]dashboards.DashboardSearchProjection, 0)
+	var page int64 = 1
+	query.SkipAccessControlFilter = true
+	// Set limit to default to prevent pagination issues
+	limit := query.Limit
+
+	for len(result) < int(limit) {
+		query.Page = page
+		query.Limit = defaultQueryLimit
+		findRes, err := dr.dashboardStore.FindDashboards(ctx, query)
+		if err != nil {
+			return nil, err
+		}
+
+		query.Limit = limit
+		res, err := dr.checkDashboards(ctx, query, findRes)
+		if err != nil {
+			return nil, err
+		}
+
+		result = append(result, res...)
+		page++
+
+		// Stop when last page reached
+		if len(findRes) < defaultQueryLimit {
+			break
+		}
+	}
+
+	if len(result) > int(limit) {
+		result = result[:limit]
+	}
+
+	return result, nil
+}
+
+func (dr *DashboardServiceImpl) checkDashboards(ctx context.Context, query *dashboards.FindPersistedDashboardsQuery, searchRes []dashboards.DashboardSearchProjection) ([]dashboards.DashboardSearchProjection, error) {
+	ctx, span := tracer.Start(ctx, "dashboards.service.checkDashboards")
+	defer span.End()
+
+	if len(searchRes) == 0 {
+		return nil, nil
+	}
+
+	orgId := query.OrgId
+	if orgId == 0 && query.SignedInUser.GetOrgID() != 0 {
+		orgId = query.SignedInUser.GetOrgID()
+	}
```

Comment:
Can we really reach a point where `orgId != signedInUser.GetOrgID()`? It would mean users are not signed in the organization they are searching dashboards in.
**If we cannot reach that point**, then I'd suggest to pass `user identity.Requester` instead of the `query` as a parameter.

Reply:
There's OrgId parameter in query itself. It allows to search in org other than user is signed in I think. 

https://github.com/grafana/grafana/blob/8e5459791bd7814f55c0bee9385af88c105195a5/pkg/services/dashboards/database/database.go#L901-L906